syntax = "proto3";

// import "google/protobuf/timestamp.proto";
//Comments

message Comment {
  int64 id = 1;
  string studentId = 2;
  string author = 9;
  string body = 3;
  int32 upvotes = 4;
  int32 downvotes = 5;
  int64 articleId = 6;
  string createdAt = 7;
  string updatedAt = 8;
}

message GetAllCommentsRequest {

}

message GetAllCommentsResponse {
  repeated Comment comments = 1;
}

message GetCommentByIdRequest {
  int64 commentId = 1;
}

message GetCommentByIdResponse {
  Comment comment = 1;
}

message SaveCommentRequest {
  Comment comment = 1;
}

message SaveCommentResponse {
  int64 commentId = 1;
  string message = 2;
}

message UpdateCommentRequest {
  Comment comment = 1;
}

message UpdateCommentResponse {
  string message = 1;
}

message DeleteCommentRequest {
  int64 commentId = 1;
}

message DeleteCommentResponse {
  string message = 1;
}

message GetAllCommentsByArticleIdRequest {
  int64 articleId = 1;
  int32 page = 2;
  int32 size = 3;
}

message GetAllCommentsByArticleIdResponse {
  repeated Comment comments = 1;
}


service CommentService {
  rpc getAllComments(GetAllCommentsRequest) returns (GetAllCommentsResponse) {}
  rpc getCommentById(GetCommentByIdRequest) returns (GetCommentByIdResponse) {}
  rpc saveComment(SaveCommentRequest) returns (SaveCommentResponse) {}
  rpc updateComment(UpdateCommentRequest) returns (UpdateCommentResponse) {}
  rpc deleteComment(DeleteCommentRequest) returns (DeleteCommentResponse) {}
  rpc getAllCommentsByArticleId(GetAllCommentsByArticleIdRequest) returns (GetAllCommentsByArticleIdResponse) {}
}

// Articles

service ArticleService {
  rpc getAllArticles(GetAllArticlesRequest) returns (GetAllArticlesResponse) {}
  rpc getArticleById(GetArticleByIdRequest) returns (GetArticleByIdResponse) {}
  rpc saveArticle(SaveArticleRequest) returns (SaveArticleResponse) {}
  rpc updateArticle(UpdateArticleRequest) returns (UpdateArticleResponse) {}
  rpc deleteArticle(DeleteArticleRequest) returns (DeleteArticleResponse) {}
}

message Article {
  int64 id = 1;
  string title = 2;
  string body = 3;
  string studentId = 4;
  string author = 11;
  string company = 5;
  int32 upvotes = 6;
  int32 downvotes = 7;
  string updatedAt = 8;
  string createdAt = 9;
  repeated Comment comments = 10;
}

message GetAllArticlesRequest {

}

message GetAllArticlesResponse {
  repeated Article articles = 1;
}

message GetArticleByIdRequest {
  int64 articleId = 1;
}

message GetArticleByIdResponse {
  Article article = 1;
}

message SaveArticleRequest {
  Article article = 1;
}

message SaveArticleResponse {
  optional int64 articleId = 1;
  string message = 2;
}

message UpdateArticleRequest {
  Article article = 1;
}

message UpdateArticleResponse {
  string message = 1;
}

message DeleteArticleRequest {
  int64 articleId = 1;
}

message DeleteArticleResponse {
  string message = 1;
}